File: golang/day15/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

func loadData(filename string) ([]string, error) {
	data, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer func() {
		err := data.Close()
		if err != nil {
			panic(err)
		}
	}()
	scanner := bufio.NewScanner(data)
	scanner.Scan()
	ops := strings.Split(scanner.Text(), ",")
	if err := scanner.Err(); err != nil {
		return nil, err
	}
	return ops, nil
}

func hash(input string) int {
	hashSum := int32(0)
	for _, c := range input {
		hashSum += c
		hashSum *= 17
		hashSum %= 256
	}
	return int(hashSum)
}

func solutionPart1() {
	ops, err := loadData("input.txt")
	if err != nil {
		panic(err)
	}
	totalSum := 0
	for _, op := range ops {
		totalSum += hash(op)
	}
	fmt.Println(totalSum)
}

type Box struct {
	indexMap     map[string]int
	values       []int
	currentIndex int
}

func newBox() Box {
	b := Box{
		indexMap:     make(map[string]int),
		values:       make([]int, 0),
		currentIndex: 0,
	}
	return b
}

func (b *Box) insert(label string, value int) {
	if index, ok := b.indexMap[label]; ok {
		b.values[index] = value
	} else {
		b.indexMap[label] = b.currentIndex
		b.values = append(b.values, value)
		b.currentIndex += 1
	}
}

func (b *Box) remove(label string) {
	if index, ok := b.indexMap[label]; ok {
		b.values[index] = -1
		delete(b.indexMap, label)
	}
}

func solutionPart2() {
	ops, err := loadData("input.txt")
	if err != nil {
		panic(err)
	}
	opMatcher, err := regexp.Compile("([a-z]+)([=-])(\\d*)")
	if err != nil {
		panic(err)
	}
	boxes := make([]Box, 256)
	for i := range boxes {
		boxes[i] = newBox()
	}
	for _, op := range ops {
		match := opMatcher.FindStringSubmatch(op)
		label := match[1]
		opType := match[2]
		if opType == "=" {
			value, err := strconv.Atoi(match[3])
			if err != nil {
				panic(err)
			}
			boxes[hash(label)].insert(label, value)
		} else {
			boxes[hash(label)].remove(label)
		}
	}
	totalSum := 0
	for index, b := range boxes {
		if b.currentIndex > 0 {
			boxSum := 0
			currentIndex := 1
			for _, v := range b.values {
				if v != -1 {
					boxSum += v * currentIndex
					currentIndex += 1
				}
			}
			totalSum += boxSum * (index + 1)
		}
	}
	fmt.Println(totalSum)
}

func main() {
	solutionPart1()
	solutionPart2()
}

File: Python/day15/main.py
import re


def load_data(filename: str) -> list[str]:
    with open(filename, encoding="utf-8") as file:
        return file.readline().split(",")


def hash_alg(input_str: str) -> int:
    hash_sum = 0
    for char in input_str:
        hash_sum += ord(char)
        hash_sum *= 17
        hash_sum %= 256
    return hash_sum


def solution_part_1():
    ops = load_data("input.txt")
    total_sum = sum([hash_alg(op) for op in ops])
    print(total_sum)


class Box:

    def __init__(self):
        self.index_map: dict[str, int] = {}
        self.values: list[int] = []
        self.current_index = 0

    def insert(self, label: str, value: int) -> None:
        if label in self.index_map:
            self.values[self.index_map[label]] = value
        else:
            self.index_map[label] = self.current_index
            self.values.append(value)
            self.current_index += 1

    def remove(self, label):
        if label in self.index_map:
            self.values[self.index_map[label]] = -1
            del self.index_map[label]


def solution_part_2():
    ops = load_data("input.txt")
    boxes = [Box() for _ in range(256)]

    op_matcher = re.compile(r"([a-z]+)([=-])(\d*)")
    for op in ops:
        match = op_matcher.match(op)
        if match:
            label = match.group(1)
            operator = match.group(2)
            if operator == "=":
                boxes[hash_alg(label)].insert(label, int(match.group(3)))
            elif operator == "-":
                boxes[hash_alg(label)].remove(label)

    total_sum = 0
    for (index, b) in enumerate(boxes):
        if b.current_index > 0:
            box_sum = 0
            current_index = 1
            for value in b.values:
                if value != -1:
                    box_sum += (value * current_index)
                    current_index += 1
            total_sum += (box_sum * (index + 1))
    print(total_sum)


def main():
    solution_part_1()
    solution_part_2()


if __name__ == "__main__":
    main()